goaviatrix/tags.go

```diff
@@ -50,81 +48,81 @@ func (c *Client) AddTags(tags *Tags) error {
-					continue
-				}
-				str := allKeys[i].String() + ":" + data.Results["usr_tags"][allKeys[i].String()]
-				tagList = append(tagList, str)
-			}
-			return tagList, nil
+	if tagsMap, ok := resp.Results["usr_tags"]; ok {
+		tags.Tags = tagsMap
+		for key, val := range tagsMap {
+			tagStr := key + ":" + val
+			tagList = append(tagList, tagStr)
 		}
 	}
 
-	return nil, nil
+	return tagList, nil
 }
 
-func (c *Client) DeleteTags(tags *Tags) error {
-	tags.CID = c.CID
-	tags.Action = "delete_resource_tag"
-	verb := "POST"
-	body := fmt.Sprintf("CID=%s&action=%s&cloud_type=%d&resource_type=%s&resource_name=%s&del_tag_list=%s",
-		c.CID, tags.Action, tags.CloudType, tags.ResourceType, tags.ResourceName, tags.TagList)
-	log.Tracef("%s %s Body: %s", verb, c.baseURL, body)
-	req, err := http.NewRequest(verb, c.baseURL, strings.NewReader(body))
-	if err == nil {
-		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
-	} else {
-		return errors.New("HTTP New Request Post delete_resource_tags failed: " + err.Error())
+func (c *Client) GetTagsMap(tags *Tags) (map[string]string, error) {
+	data := map[string]string{
+		"action":        "list_resource_tags",
+		"CID":           c.CID,
+		"cloud_type":    strconv.Itoa(tags.CloudType),
+		"resource_type": tags.ResourceType,
+		"resource_name": tags.ResourceName,
 	}
-	resp, err := c.HTTPClient.Do(req)
+
+	var resp TagAPIResp
+	err := c.GetAPI(&resp, data["Action"], data, BasicCheck)
 	if err != nil {
-		return errors.New("HTTP Post delete_resource_tags failed: " + err.Error())
+		return nil, fmt.Errorf("HTTP Get list_resource_tags failed: %v", err)
 	}
-	var data APIResp
-	buf := new(bytes.Buffer)
-	buf.ReadFrom(resp.Body)
-	bodyString := buf.String()
-	bodyIoCopy := strings.NewReader(bodyString)
-	if err = json.NewDecoder(bodyIoCopy).Decode(&data); err != nil {
-		return errors.New("Json Decode delete_resource_tags failed: " + err.Error() + "\n Body: " + bodyString)
+
+	if !resp.Return {
+		return nil, fmt.Errorf("rest API list_resource_tags Post failed: %s", resp.Reason)
 	}
-	if !data.Return {
-		return errors.New("Rest API delete_resource_tags Post failed: " + data.Reason)
+
+	if tagsMap, ok := resp.Results["usr_tags"]; ok {
+		return tagsMap, nil
 	}
-	return nil
+	return nil, nil
 }
 
-func (c *Client) AzureUpdateTags(tags *Tags) error {
+func (c *Client) DeleteTags(tags *Tags) error {
+	params := map[string]string{
+		"action":        "delete_resource_tag",
+		"CID":           c.CID,
+		"cloud_type":    strconv.Itoa(tags.CloudType),
+		"del_tag_list":  tags.TagList,
+		"resource_name": tags.ResourceName,
+		"resource_type": tags.ResourceType,
+	}
+
+	return c.PostAPI(params["action"], params, BasicCheck)
+}
+
+func (c *Client) UpdateTags(tags *Tags) error {
 	tags.CID = c.CID
 	tags.Action = "update_resource_tags"
 
 	return c.PostAPI(tags.Action, tags, BasicCheck)
 }
+
+func TagsMapToString(tagsMap map[string]string) string {
```

Comment:
Since this helper function is only used in the `aviatrix` package perhaps we should move it there